lib/html5ever/precompiled.ex

```diff
@@ -0,0 +1,303 @@
+defmodule Html5ever.Precompiled do
+  @moduledoc false
+
+  require Logger
+
+  @available_targets ~w(
+    aarch64-apple-darwin
+    x86_64-apple-darwin
+    x86_64-unknown-linux-gnu
+    x86_64-unknown-linux-musl
+    arm-unknown-linux-gnueabihf
+    x86_64-pc-windows-msvc
+    x86_64-pc-windows-gnu
+  )
+  @available_nif_versions ~w(2.14 2.15 2.16)
+
+  @doc """
+  Returns the target triple for download or compile and load.
+
+  This function is translating and adding more info to the system
+  architecture returned by Elixir/Erlang to one used by Rust.
+
+  The returning string format is the following:
+
+    "nif-NIF_VERSION-ARCHITECTURE-VENDOR-OS-ABI"
+
+  ## Examples
+
+      iex> target()
+      {:ok, "nif-2.16-x86_64-unknown-linux-gnu"} 
+
+      iex> target()
+      {:ok, "nif-2.15-aarch64-apple-darwin"}
+
+  """
+  def target(config \\ target_config()) do
+    sys_arch = maybe_override_with_env_vars(config.system_arch)
+
+    arch_os =
+      case config.os_type do
+        {:unix, os} ->
+          os
+          |> normalize_arch_os(sys_arch)
+          |> system_arch_to_string()
+
+        {:win32, _} ->
+          # 32 or 64 bits
+          arch =
+            case config.word_size do
+              4 -> "i686"
+              8 -> "x86_64"
+              _ -> "unknown"
+            end
+
+          sys_arch
+          |> Map.put_new(:arch, arch)
+          |> Map.put_new(:vendor, "pc")
+          |> Map.put_new(:os, "windows")
+          |> Map.put_new(:abi, "msvc")
+          |> system_arch_to_string()
+      end
+
+    cond do
+      arch_os not in @available_targets ->
+        {:error,
+         "precompiled NIF is not available for this target: #{inspect(arch_os)}. The available targets are:\n - #{Enum.join(@available_targets, "\n - ")}"}
+
+      config.nif_version not in @available_nif_versions ->
+        {:error,
+         "precompiled NIF is not available for this NIF version: #{inspect(config.nif_version)}. The available NIF versions are:\n - #{Enum.join(@available_nif_versions, "\n - ")}"}
+
+      true ->
+        {:ok, "nif-#{config.nif_version}-#{arch_os}"}
+    end
+  end
+
+  defp target_config do
+    current_nif_version = :erlang.system_info(:nif_version) |> List.to_string()
+
+    %{
+      os_type: :os.type(),
+      system_arch: system_arch(),
+      word_size: :erlang.system_info(:wordsize),
+      nif_version: find_compatible_nif_version(current_nif_version, @available_nif_versions)
+    }
+  end
+
+  # In case one is using this lib in a newer OTP version, we try to
+  # find the latest compatible NIF version.
+  defp find_compatible_nif_version(vsn, available) do
+    if vsn in available do
+      vsn
+    else
+      [major | _] = String.split(vsn, ".")
+
+      available
+      |> Enum.filter(fn vsn -> String.starts_with?(vsn, major <> ".") end)
+      |> Enum.max_by(fn vsn -> vsn |> String.split(".") |> Enum.map(&String.to_integer/1) end)
```

Comment:
Is this missing a filter for (minor version) <= (minor version of current BEAM)?

Reply:
@philss I think we need something like this:

```
  defp find_compatible_nif_version(vsn, available) do
    if vsn in available do
      {:ok, vsn}
    else
      [major, minor | _] = parse_version(vsn)

      available
      |> Enum.map(&parse_version/1)
      |> Enum.filter(fn
        [^major, available_minor | _] when available_minor >= minor -> true
        [_ | _] -> false
      end)
      |> case do
        [] -> :error
        match -> {:ok, match |> Enum.max() |> Enum.join(".")}
      end
    end
  end

  defp parse_version(vsn) do
    vsn |> String.split(".") |> Enum.map(&String.to_integer/1)
  end
```

We should expose this function (with `@doc false` or on a private module) and write some unit tests.